### DISCUSSION TITLE
Named routes?

### DISCUSSION THREAD

**Discussion by vincent-thomas** [TanStack/router#206] [Ideas]
## Named Routes?

Hi! Is it possible to have named routes? It would be really nice to have it implemented (if its already not) like the angular way, where you set an key to the <router-outlet> and then put a key in the router object. Maybe this is already implemented, but i couldn't find anything in the docs, could someone please point me in the right direction?

**Comment by tom-sherman** [comment]
You can give the route an id, if that's what you mean?

The route's name could also conceptually be it's path relative to it's position in the router hierarchy, react-location surfaces this to you as the history pathName,

**Reply by vincent-thomas** [reply]
Its hard to describe with words i'm going to try to explain it to you with a example

You have a root component, lets call that Root, so in jsx this would be <Root />

In the `Root` component you would have something like this

```jsx
....
  return (
    <>
      <Outlet name="first"/>
      <Outlet name="second" />

    </>
  )
```

The router would look something like this.

```jsx
{
  path: "/",
  component: <Root />
  children: [
    {
      name: "first",
      component: <h1>first</h1>
    },
    {
      name: "first",
      component: <h1>second</h1>
    }
  ]
}
```

With named routes that would render this

```jsx
<h1>first</h1>
<h1>second</h1>
```

on the "/" page. Does react-location behave in this way with `id`? And is there a way to make this work (or in a similay way?) with the current state of the project?

**Reply by vincent-thomas** [reply]
> You can give the route an id, if that's what you mean?
> 
> The route's name could also conceptually be it's path relative to it's position in the router hierarchy, react-location surfaces this to you as the history pathName,

Does id-field do anything?

**Comment by log1st** [comment]
You can use Router Context:

```typescript
// rootRoute.ts
export type AppRouterContext = {
  sidebar?: FC;
};

export const rootRoute = createRootRouteWithContext<AppRouterContext>()({
  component: RootLayout,
});
```

```typescript
// MainLayout.tsx

function MainLayout() {
  const route = useRouterState();

  const { sidebar: Sidebar } = useRouteContext({
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    from: route.location.href as any,
  });
  
  return (
    <div>
      {Sidebar ? <div class="sidebar"><Sidebar/></div> : null}
      <Outlet/> 
    </div>
  )
}
```

```typescript
// usersRoute.ts
export const usersRoute = createRoute({
  getParentRoute: () => rootRoute,
  path: "users",
  component: UsersPage,
  context: () => ({
    sidebar: UsersSidebar
  })
});
```

like so.
